lightning/src/ln/peer_handler.rs

```diff
@@ -450,26 +463,55 @@ impl<Descriptor: SocketDescriptor, CM: Deref> PeerManager<Descriptor, CM> where
 				Some(peer) => {
-					assert!(peer.pending_read_buffer.len() > 0);
-					assert!(peer.pending_read_buffer.len() > peer.pending_read_buffer_pos);
+					peer.pending_read_buffer.extend_from_slice(&data);
+					while peer.pending_read_buffer.len() > 0 {
 
-					let mut read_pos = 0;
-					while read_pos < data.len() {
-						{
-							let data_to_copy = cmp::min(peer.pending_read_buffer.len() - peer.pending_read_buffer_pos, data.len() - read_pos);
-							peer.pending_read_buffer[peer.pending_read_buffer_pos..peer.pending_read_buffer_pos + data_to_copy].copy_from_slice(&data[read_pos..read_pos + data_to_copy]);
-							read_pos += data_to_copy;
-							peer.pending_read_buffer_pos += data_to_copy;
+						let mut conduit_option = None;
+						let mut remote_pubkey_option = None;
+
+						if let &mut PeerState::Handshake(ref mut handshake) = &mut peer.encryptor {
+							let (response, conduit, remote_pubkey) = handshake.process_act(&peer.pending_read_buffer, None).unwrap();
+							peer.pending_read_buffer.drain(..); // we read it all
+
+							if let Some(key) = remote_pubkey {
+								remote_pubkey_option = Some(key);
+							}
+
+							peer.pending_outbound_buffer.push_back(response);
+							if let Some(conduit) = conduit {
+								conduit_option = Some(conduit);
+							}
 						}
 
-						if peer.pending_read_buffer_pos == peer.pending_read_buffer.len() {
-							peer.pending_read_buffer_pos = 0;
+						if let Some(key) = remote_pubkey_option {
+							peer.their_node_id = Some(key);
+						}
+
+						if let Some(conduit) = conduit_option {
+							// Rust 1.22 does not allow assignment in a borrowed context, even if mutable
+							peer.encryptor = PeerState::Connected(conduit);
+
+							// the handshake has finished, so one way or another, we now have their node id
+							match peers.node_id_to_descriptor.entry(peer.their_node_id.unwrap()) {
+								hash_map::Entry::Occupied(_) => {
+									log_trace!(self, "Got second connection with {}, closing", log_pubkey!(peer.their_node_id.unwrap()));
+									peer.their_node_id = None; // Unset so that we don't generate a peer_disconnected event
+									return Err(PeerHandleError{ no_connection_possible: false })
+								},
+								hash_map::Entry::Vacant(entry) => {
+									log_trace!(self, "Finished noise handshake for connection with {}", log_pubkey!(peer.their_node_id.unwrap()));
+									entry.insert(peer_descriptor.clone())
+								},
+							};
+						}
 
+						if let &mut PeerState::Connected(ref mut conduit) = &mut peer.encryptor {
 							macro_rules! encode_and_send_msg {
 								($msg: expr) => {
 									{
 										log_trace!(self, "Encoding and sending message of type {} to {}", $msg.type_id(), log_pubkey!(peer.their_node_id.unwrap()));
-										peer.pending_outbound_buffer.push_back(peer.channel_encryptor.encrypt_message(&encode_msg!(&$msg)[..]));
+										// we are in a context where conduit is known
```

Comment:
I think it doesn't make sense because the other variables are assumed to be known, too. Once this is refactored from a macro into a regular method, it will solve itself.